lib/mix/lib/mix/utils.ex

```diff
@@ -393,12 +391,54 @@ defmodule Mix.Utils do
     URI.parse(path).scheme in ["http", "https"]
   end
 
-  defp proxy(proxy_scheme, proxy) do
-    uri  = URI.parse(proxy)
+  def proxy_config(url) do
+    {http_proxy, https_proxy} = proxy_env
+
+    proxy_auth(URI.parse(url), http_proxy, https_proxy)
+  end
+
+  defp proxy_env do
+    http_proxy  = System.get_env("HTTP_PROXY")  || System.get_env("http_proxy")
+    https_proxy = System.get_env("HTTPS_PROXY") || System.get_env("https_proxy")
+
+    {proxy_setup(:http, http_proxy), proxy_setup(:https, https_proxy)}
+  end
+
+  defp proxy_setup(scheme, proxy) do
+    uri = URI.parse(proxy || "")
 
     if uri.host && uri.port do
       host = String.to_char_list(uri.host)
-      :httpc.set_options([{proxy_scheme, {{host, uri.port}, []}}], :mix)
+      :httpc.set_options([{proxy_scheme(scheme), {{host, uri.port}, []}}], :mix)
     end
+
+    uri
+  end
+
+  defp proxy_scheme(scheme) do
+    case scheme do
+      :http  -> :proxy
+      :https -> :https_proxy
+    end
+  end
+
+  defp proxy_auth(%URI{scheme: "http"}, http_proxy, _https_proxy),
+    do: proxy_auth(http_proxy)
+  defp proxy_auth(%URI{scheme: "https"}, _http_proxy, https_proxy),
+    do: proxy_auth(https_proxy)
+
+  defp proxy_auth(nil),
+    do: []
+  defp proxy_auth(%URI{userinfo: nil}),
+    do: []
+  defp proxy_auth(%URI{userinfo: auth}) do
+    [user, pass] = case auth |> String.split(":") do
```

Comment:
Don't pipe single arguments like this. Write `String.split(auth, ":")` instead, it is clearer.